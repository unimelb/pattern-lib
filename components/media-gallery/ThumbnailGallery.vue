<template>
  <div class="thumbnails">
    <div
      :class="thumbnailColumnClass"
      class="thumbnails__container">
      <div
        v-for="(item, index) in items"
        :key="index"
        :class="thumbnailClass"
        class="thumbnails__item">
        <div
          :class="thumbnailOpenClass"
          tabindex="0"
          @click="useOverlay && callback(index)"
          @keydown.enter="callback(index)">
          <div
            v-if="item.type === 'image'"
            :aria-label="item.altText"
            :style="{ backgroundImage: `url(${item.src})` }"
            class="thumbnails__image" />

          <div
            v-if="item.type === 'video'"
            class="thumbnails__video">
            <VideoEmbed
              :src="item.src"
              :class="embedClass"
              class="thumbnails__video-embed" />
          </div>
        </div>

        <div v-if="displayCaption">
          <div class="thumbnails__caption">
            {{ item.title }}
          </div>
          <p class="thumbnails__description">
            {{ item.description }}
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import VideoEmbed from '../embed/VideoEmbed.vue';

export default {
  components: {
    VideoEmbed,
  },
  props: {
    columns: {
      type: String,
      default: '1',
    },
    items: {
      type: Array,
      default: () => [{}],
    },
    callback: {
      type: Function,
      default: () => {},
    },
    displayCaption: {
      type: Boolean,
      default: false,
    },
    useOverlay: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    thumbnailColumnClass() {
      const { columns } = this;

      return {
        [`thumbnails__container--${columns}-col`]: true,
      };
    },
    thumbnailClass() {
      return {
        'thumbnails__item--no-caption': !this.displayCaption,
      };
    },
    thumbnailOpenClass() {
      return {
        'thumbnails__open-overlay': this.useOverlay,
      };
    },
    embedClass() {
      return {
        'thumbnails__video-embed--overlay': this.useOverlay,
      };
    },
  },
};
</script>
