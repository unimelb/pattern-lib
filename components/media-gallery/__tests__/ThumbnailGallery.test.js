import { shallow, mount } from 'vue-test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import sinon from 'sinon';
import ThumbnailGallery from '../ThumbnailGallery.vue';

expect.extend(toHaveNoViolations);

const itemsMock = [
  {
    src: 'test',
    title: 'test1',
    description: 'test1',
    type: 'image',
  },
  {
    src: 'test2',
    title: 'test2',
    description: 'test2',
    type: 'video',
  },
];

describe('ThumbnailGallery', () => {
  it('should match snapshot', () => {
    const result = shallow(ThumbnailGallery).element;
    expect(result).toMatchSnapshot();
  });

  it('should have default props and correct types images/callback/columns/displayCaption/overlay', () => {
    const wrapper = shallow(ThumbnailGallery);
    const {
      items,
      callback,
      columns,
      displayCaption,
      useOverlay,
    } = wrapper.vm.$options.props;

    expect(columns.type).toBe(String);
    expect(wrapper.props().columns).toBe('1');
    expect(items.type).toBe(Array);
    expect(callback.type).toBe(Function);
    expect(displayCaption.type).toBe(Boolean);
    expect(useOverlay.type).toBe(Boolean);
    expect(wrapper.props().items).toEqual([{}]);
  });

  it('should accept columns prop with correct type', () => {
    const columns = '3';
    const wrapper = shallow(ThumbnailGallery, {
      propsData: {
        columns,
      },
    });

    expect(typeof wrapper.props().columns).toBe('string');
    expect(wrapper.props().columns).toBe(columns);
  });

  it('should render images from prop', () => {
    const wrapper = mount(ThumbnailGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    expect(wrapper.props().items.length).toBe(2);
    expect(wrapper.find('.thumbnails__image').attributes().style).toBe('background-image: url(test);');
  });

  it('should render video from prop', () => {
    const wrapper = mount(ThumbnailGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    expect(wrapper.props().items.length).toBe(2);
    expect(wrapper.find('.embed').exists()).toBe(true);
  });

  it('should render caption from prop', () => {
    const wrapper = mount(ThumbnailGallery, {
      propsData: {
        displayCaption: true,
      },
    });
    expect(wrapper.props().displayCaption).toBe(true);
    expect(wrapper.find('.thumbnails__description').exists()).toBe(true);
  });

  it('should have click event on thumb', () => {
    const callback = sinon.spy();
    const wrapper = mount(ThumbnailGallery, {
      propsData: {
        items: itemsMock,
        callback,
        useOverlay: true,
      },
    });
    wrapper.find('.thumbnails__item div').trigger('click');
    expect(callback.called).toBe(true);
  });

  it('Component throws no accessibility violations', (done) => {
    const html = shallow(ThumbnailGallery).html();
    // pass anything that outputs html to axe
    return axe(html).then((response) => {
      expect(response).toHaveNoViolations();
      done();
    });
  });
});
