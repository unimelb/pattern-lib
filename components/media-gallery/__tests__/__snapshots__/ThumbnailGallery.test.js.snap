// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThumbnailGallery should match snapshot 1`] = `
<div
  class="thumbnails"
>
  <div
    class="thumbnails__container thumbnails__container--1-col"
  >
    <div
      class="thumbnails__item thumbnails__item--no-caption"
    >
      <div
        class=""
        tabindex="0"
      >
        <!---->
         
        <!---->
      </div>
       
      <!---->
    </div>
  </div>
</div>
`;
